### PR TITLE
CA-297086: Improve the function that uploads a server status report, …

### PR DESCRIPTION
**Pull request by MihaelaStoica**
…to avoit timeout exceptions

The changes included are:
- use a HttpClient object instead of HttpWebRequest
- use the same FileStream for all the chunk uploads
- retry the chunk upload if it fails (max three tries)

Signed-off-by: Mihaela Stoica <mihaela.stoica@citrix.com>

### CODE REVIEW

File: XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs [hunk 1]
```diff
@@ -175,62 +159,60 @@ public string UploadZip(string fileName, string caseNumber, System.Threading.Can
...
-                            // This chunk is successfully uploaded
-                            offset += thisChunkSize;
-                            break;
-                        }
-
-                        // Fail to upload the chunk for 3 times so fail the bundle upload.
-                        if (i == 2)
+                    try
+                    {
+                        for (int i = 0; i < 3; i++)
```
Comment (kc284) [inline XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs]: Let's put 3 in a constant and use it further down, so as to have one place to change if we update the number of retries.

File: XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs [hunk 2]
```diff
@@ -214,5 +229,12 @@ public string UploadZip(string fileName, string caseNumber, System.Threading.Can
             log.InfoFormat("Succeeded to upload bundle {0}", fileName);
             return uploadUuid;
         }
+
+        #region IDisposable Members
+        public void Dispose()
+        {
+            httpClient.Dispose();
```
Comment (kc284) [inline XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs]: The HttpClientHandler is missing before httpClient (it will have to be promoted from local variable to class field).

File: XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs [hunk 3]
```diff
@@ -132,23 +139,31 @@ private bool UploadChunk(string address, FileStream source, long offset, long th
...
+                        if (response.IsSuccessStatusCode)
+                        {
+                            var respString = response.Content.ReadAsStringAsync().Result;
+
+                            // After sending every chunk upload request to server, response will contain a status indicating if it is complete.
+                            var res = (Dictionary<string, object>) serializer.DeserializeObject(respString);
+                            log.InfoFormat("The status of chunk upload: {0}", res.ContainsKey("status") ? res["status"] : "");
+                            return true;
+                        }
+                        log.ErrorFormat("Failed to upload the chunk. The exception was: {0} ({1})", response.StatusCode, response.ReasonPhrase);
```
Comment (kc284) [inline XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs]: Strictly speaking this is not an exception, I'd rephrase the message to `Failed to upload the chunk. POST request returned {0} ({1})`.

File: XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs [hunk 4]
```diff
@@ -80,34 +80,41 @@ public XenServerHealthCheckUpload(string token, int verbosity, string uploadUrl,
...
+                        if (response.IsSuccessStatusCode)
+                        {
+                            var respString = response.Content.ReadAsStringAsync().Result;
+
+                            // Get the upload uuid
+                            var res = (Dictionary<string, object>) serializer.DeserializeObject(respString);
+                            if (res.ContainsKey("id"))
+                                return (string) res["id"];
+                        }
+                        log.ErrorFormat("Exception while initiating a new CIS upload. The exception was: {0} ({1})", response.StatusCode, response.ReasonPhrase);
```
Comment (kc284) [inline XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs]: Strictly speaking this is not an exception. I'd rephrase the message to `Failed to initiate a CIS upload. POST request returned {0} ({1)}`.
Also, not sure if it is possible, but if the status code is 200 and the response does not contain the id, the error will still be logged, which is misleading. We can either put the log in an else block, or preferably (to match the upload chunk case) do `return res.ContainsKey("id") ? (string) res["id"] : "";`

File: XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs (hunk 2 above)
Comment (MihaelaStoica) [reply XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs]: The HttpClientHandler is disposed automatically by HttpClient.Dispose(), because the HttpClient constructor used, sets the internal flag `disposeHandler` to true by default. See the constructor implementation in the HttpClient class:
``` 
    /// <summary>Initializes a new instance of the <see cref="T:System.Net.Http.HttpClient" /> class with a specific handler.</summary>
    /// <param name="handler">The HTTP handler stack to use for sending requests. </param>
    [__DynamicallyInvokable]
    public HttpClient(HttpMessageHandler handler)
      : this(handler, true)
    {
    }

    /// <summary>Initializes a new instance of the <see cref="T:System.Net.Http.HttpClient" /> class with a specific handler.</summary>
    /// <param name="handler">The <see cref="T:System.Net.Http.HttpMessageHandler" /> responsible for processing the HTTP response messages.</param>
    /// <param name="disposeHandler">
    /// <see langword="true" /> if the inner handler should be disposed of by Dispose(), <see langword="false" /> if you intend to reuse the inner handler.</param>
    [__DynamicallyInvokable]
    public HttpClient(HttpMessageHandler handler, bool disposeHandler)
      : base(handler, disposeHandler)
    {
    ....
```
However, as this is not clear in the code, I will change it to explicitily use the second constructor, `HttpClient(HttpMessageHandler handler, bool disposeHandler)`.

Comment (kc284) [reply XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs]: Fair enough.
